kyo-stm/shared/src/main/scala/kyo/STM.scala

```diff
@@ -108,34 +111,97 @@ object STM:
-                                    locked.foreach((ref, entry) => ref.unlock(entry))
-                                    Abort.fail(FailedTransaction(frame))
-                                else
-                                    // Successfully locked all references - commit changes
-                                    locked.foreach((ref, entry) => ref.commit(tid, entry))
-                                    // Release all locks
-                                    locked.foreach((ref, entry) => ref.unlock(entry))
+                        Var.runWith(TRefLog.empty)(v) { (log, result) =>
+                            val logMap = log.toMap
+                            logMap.size match
+                                case 0 =>
+                                    // Nothing to commit
                                     result
-                                end if
-                            }
+                                case 1 =>
+                                    // Fast-path for a single ref
+                                    IO.Unsafe {
+                                        val (ref, entry) = logMap.head
+                                        // No need to pre-validate since `lock` validates and
+                                        // there's a single ref
+                                        if ref.lock(entry) then
+                                            ref.commit(tid, entry)
+                                            ref.unlock(entry)
+                                            result
+                                        else
+                                            Abort.fail(FailedTransaction(frame))
+                                        end if
+                                    }
+                                case size =>
+                                    // Commit multiple refs
+                                    IO.Unsafe {
+                                        // Flattened representation of the log
+                                        val array = new Array[Any](size * 2)
+
+                                        try
+                                            def fail = throw new FailedTransaction(frame)
+
+                                            var i = 0
+                                            // Pre-validate and dump the log to the flat array
+                                            logMap.foreachEntry { (ref, entry) =>
+                                                // This code uses exception throwing because
+                                                // foreachEntry is the only way to traverse the
+                                                // map without allocating tuples, so throwing
+                                                // is the workaround to short circuit
+                                                if !ref.validate(entry) then fail
```

Comment:
I've added an initial validation of the ref log first so it's possible to fail invalid transactions early without locking any refs. It has some performance cost but helps reduce the impact of retries.